easybuild/framework/easystack.py

```diff
@@ -0,0 +1,243 @@
+from easybuild.base import fancylogger
+try:
+    import yaml
+except ImportError:
+    pass
+_log = fancylogger.getLogger('easystack', fname=False)
+
+
+class Easystack(object):
+    """One class instance per easystack. General options + list of all SoftwareSpecs instances"""
+    def __init__(self):
+        self.easybuild_version = None
+        self.robot = False
+        self.software_list = []
+
+    def compose_ec_names(self):
+        """Returns a list of all easyconfig names"""
+        ec_filenames = []
+        for sw in self.software_list:
+            full_ec_version = self.det_full_ec_version({
+                'toolchain': {'name': sw.toolchain_name, 'version': sw.toolchain_version},
+                'version': sw.version,
+                'versionsuffix': sw.versionsuffix,
+            })
+            ec_filename = '%s-%s.eb' % (sw.name, full_ec_version)
+            ec_filenames.append(ec_filename)
+        return ec_filenames
+
+    def det_full_ec_version(self, version_specs):
+        ec_version_string = '%s-%s-%s%s' % (
+            version_specs['version'], version_specs['toolchain']['name'],
+            version_specs['toolchain']['version'], version_specs['versionsuffix']
+        )
+        return ec_version_string
+
+    # flags applicable to all sw (i.e. robot)
+    def get_general_options(self):
+        """Returns general options (flags applicable to all sw (i.e. --robot))"""
+        general_options = {}
+        # TODO add support for general_options
+        # general_options['robot'] = self.robot
+        # general_options['easybuild_version'] = self.easybuild_version
+        return general_options
+
+
+class SoftwareSpecs(object):
+    """Contains information about every SW that should be installed"""
+    def __init__(self, name, version, versionsuffix, toolchain, toolchain_version, toolchain_name):
+        self.name = name
+        self.version = version
+        self.toolchain = toolchain
+        self.toolchain_version = toolchain_version
+        self.toolchain_name = toolchain_name
+        self.toolchain = toolchain
+        self.versionsuffix = versionsuffix
+
+
+# implement this to your own needs - to create custom yaml/json/xml parser
+class GenericEasystackParser(object):
+    """Parent of all implemented parser classes"""
+    @ staticmethod
+    def parse(filepath):
+        """Parent of all implemented parser functions (i.e. JSON, XML...)"""
+        raise NotImplementedError
+
+
+class YamlEasystackParser(GenericEasystackParser):
+    """YAML file parser"""
+    @ staticmethod
+    def parse(filepath):
+        """Parses YAML file and assigns obtained values to SW config instances as well as general config instance"""
+        try:
+            with open(filepath, 'r') as f:
+                spec_dict = yaml.safe_load(f)
+
+                eb = Easystack()
+        except Exception:
+            raise EasyBuildError("Could not read provided easystack.")
+
+        try:
```

Comment:
This is a huge `try` block. PEP recommends keeping try blocks as short as possible "Additionally, for all try/except clauses, limit the try clause to the absolute minimum amount of code necessary. Again, this avoids masking bugs." https://legacy.python.org/dev/peps/pep-0008/#programming-recommendations 

I agree, since it is hard to see what failure mode this `try` is testing for, and the `except` can thus only be very generic.

I guess here, you just want to `try` if the `software` keyword is in the spec_dict? That's what I would limit myself too then, so you can throw a super clear error ("Not a valid EasyStack YAML file: no 'software' key found" or something of that nature)

Reply:
you are right, it has been there more or less because I was hoping to avoid doing try-except for all possible index errors that could occur and have one for all missing keywords...but i've already dived into it, so it does make sense to finish for all...